# __Import Library__

In [1]:
# Pengolahan Data
import numpy as np
import pandas as pd

# ___Data Understanding___

### __A. _Read Data___

Pertama import terlebih dahulu dataset `transjakarta.csv` dan tampilkan informasi mengenai dataset tersebut.

In [2]:
df = pd.read_csv('Transjakarta.csv', parse_dates=['tapInTime', 'tapOutTime'])
df.head()

transID         payCardID payCardBank      payCardName payCardSex  \
0  EIIW227B8L34VB   180062659848800      emoney   Bajragin Usada          M   
1  LGXO740D2N47GZ  4885331907664776         dki     Gandi Widodo          F   
2  DJWR385V2U57TO  4996225095064169         dki    Emong Wastuti          F   
3  JTUZ800U7C86EH      639099174703       flazz     Surya Wacana          F   
4  VMLO535V7F95NJ      570928206772       flazz  Embuh Mardhiyah          M   

   payCardBirthDate corridorID                              corridorName  \
0              2008          5                     Matraman Baru - Ancol   
1              1997         6C  Stasiun Tebet - Karet via Patra Kuningan   
2              1992        R1A                        Pantai Maju - Kota   
3              1978        11D       Pulo Gebang - Pulo Gadung 2 via PIK   
4              1982         12                     Tanjung Priok - Pluit   

   direction tapInStops  ... tapInStopsLon  stopStartSeq           tapInTime  \
0        1.0     P00142  ...     106.84402             7 2023-04-03 05:21:44   
1        0.0    B01963P  ...     106.83302            13 2023-04-03 05:42:44   
2        0.0    B00499P  ...     106.81435            38 2023-04-03 05:59:06   
3        0.0    B05587P  ...     106.93526            23 2023-04-03 05:44:51   
4        0.0     P00239  ...     106.88900             5 2023-04-03 06:17:35   

   tapOutStops        tapOutStopsName tapOutStopsLat tapOutStopsLon  \
0       P00253                Tegalan      -6.203101      106.85715   
1      B03307P    Sampoerna Strategic      -6.217152      106.81892   
2      B04962P  Simpang Kunir Kemukus      -6.133731      106.81475   
3      B03090P      Raya Penggilingan      -6.183068      106.93194   
4       P00098       Kali Besar Barat      -6.135355      106.81143   

   stopEndSeq          tapOutTime  payAmount  
0        12.0 2023-04-03 06:00:53     3500.0  
1        21.0 2023-04-03 06:40:01     3500.0  
2        39.0 2023-04-03 06:50:55     3500.0  
3        29.0 2023-04-03 06:28:16     3500.0  
4        15.0 2023-04-03 06:57:03     3500.0  

[5 rows x 22 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37900 entries, 0 to 37899
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transID           37900 non-null  object        
 1   payCardID         37900 non-null  int64         
 2   payCardBank       37900 non-null  object        
 3   payCardName       37900 non-null  object        
 4   payCardSex        37900 non-null  object        
 5   payCardBirthDate  37900 non-null  int64         
 6   corridorID        36643 non-null  object        
 7   corridorName      35970 non-null  object        
 8   direction         37900 non-null  float64       
 9   tapInStops        36687 non-null  object        
 10  tapInStopsName    37900 non-null  object        
 11  tapInStopsLat     37900 non-null  float64       
 12  tapInStopsLon     37900 non-null  float64       
 13  stopStartSeq      37900 non-null  int64         
 14  tapInTime         3790

Sekilas dari informasi di atas didapat banyak nilai null dari beberapa kolom pada dataset `transjakarta.csv`. Observasi lebih lanjut diperlukan untuk analisis karakteristik dari dataset tersebut, observasi nilai null akan dilakukan di bagian _Data Cleaning_ untuk mengecek apa saja tipe _missing value_-nya untuk selanjutnya dilakukan _handling_.

### __B. Identifikasi Baris & Kolom__

In [4]:
display(df.shape, df.columns)

(37900, 22)

Index(['transID', 'payCardID', 'payCardBank', 'payCardName', 'payCardSex',
       'payCardBirthDate', 'corridorID', 'corridorName', 'direction',
       'tapInStops', 'tapInStopsName', 'tapInStopsLat', 'tapInStopsLon',
       'stopStartSeq', 'tapInTime', 'tapOutStops', 'tapOutStopsName',
       'tapOutStopsLat', 'tapOutStopsLon', 'stopEndSeq', 'tapOutTime',
       'payAmount'],
      dtype='object')

> Terdapat 37900 Baris dan 22 Kolom

> Karena nama-nama kolom yang tertera panjang dan cukup sulit untuk dibedakan, maka dilakukan _rename_ agar memudahkan dalam analisisnya

In [5]:
df.rename(columns={'transID':'TransID',
                    'payCardID':'CardID',
                    'payCardBank':'PayCardBank',
                    'payCardName':'Name',
                    'payCardSex':'Gender',
                    'payCardBirthDate': 'BirthYear',
                    'corridorID':'CorridorID',
                    'corridorName':'CorridorName',
                    'direction':'Direction',
                    'tapInStops':'TapIn_ID',
                    'tapInStopsName':'TapIn_Name',
                    'tapInStopsLat':'TapIn_Lat',
                    'tapInStopsLon':'TapIn_Lon',
                    'stopStartSeq':'TapIn_Seq',
                    'tapInTime':'TapIn_Time',
                    'tapOutStops':'TapOut_ID',
                    'tapOutStopsName':'TapOut_Name',
                    'tapOutStopsLat':'TapOut_Lat',
                    'tapOutStopsLon':'TapOut_Lon',
                    'stopEndSeq':'TapOut_Seq',
                    'tapOutTime':'TapOut_Time',
                    'payAmount':'PayAmount'
                    }, inplace=True)

pd.set_option('display.max_columns', None)
df.head(1)

TransID           CardID PayCardBank            Name Gender  \
0  EIIW227B8L34VB  180062659848800      emoney  Bajragin Usada      M   

   BirthYear CorridorID           CorridorName  Direction TapIn_ID TapIn_Name  \
0       2008          5  Matraman Baru - Ancol        1.0   P00142  Pal Putih   

   TapIn_Lat  TapIn_Lon  TapIn_Seq          TapIn_Time TapOut_ID TapOut_Name  \
0  -6.184631  106.84402          7 2023-04-03 05:21:44    P00253     Tegalan   

   TapOut_Lat  TapOut_Lon  TapOut_Seq         TapOut_Time  PayAmount  
0   -6.203101   106.85715        12.0 2023-04-03 06:00:53     3500.0

__Deskripsi Kolom__

Nama Kolom | _Replacement_ Nama Kolom | Dtype | Deskripsi |
| --- | --- | --- | --- |
transID|`TransID`| object | ID Transaksi yang __unik__ di setiap transaksi
payCardID|`CardID`| int64 |Pengidentifikasi utama _Customer_. Kartu tersebut digunakan _Customer_ sebagai tiket masuk dan keluar.
payCardBank|`PayCardBank`| object | __Nama bank__ penerbit kartu _Customer_
payCardName|`Name`| object | __Nama Customer__ yang tercantum di kartu.
payCardSex|`Gender`| object | __Jenis kelamin Customer__ yang tercantum di kartu.
payCardBirthDate|`BirthYear`| int64 | __Tahun lahir Customer__ yang tercantum di kartu.
corridorID|`CorridorID`| object | ID Koridor / __ID Rute__ sebagai kunci pengelompokan rute.
corridorName|`CorridorName`| object | Nama Koridor / __Nama Rute__ berisi __'awal'__ dan __'akhir'__ untuk setiap rute.
direction|`Direction`| float64 | __Arah rute__. __0__ untuk arah __'Pergi'__, __1__ untuk arah __'Kembali'__. 
tapInStops|`TapIn_ID`| object | __ID Pintu Masuk__ (perhentian masuk), untuk mengidentifikasi nama perhentian.
tapInStopsName|`TapIn_Name`| object | __Nama pintu masuk__ tempat _Customer_ _tap-in_ kartu.
tapInStopsLat|`TapIn_Lat`| float64 | __Garis Lintang__ _tap-in_ kartu di pintu masuk
tapInStopsLon|`TapIn_Lon`| float64 | __Garis Bujur__ _tap-in_ kartu di pintu masuk
stopStartSeq|`TapIn_Seq`| int64 | __Urutan pemberhentian__, pemberhentian pertama, pemberhentian kedua, dst. Terkait dengan arah.
tapInTime|`TapIn_Time`| datetime64[ns] | __Waktu__ ketika _Customer_ masuk (_tap-in_ kartu). __Tanggal dan detail waktu__.
tapOutStops|`TapOut_ID`| object | __ID Pintu Keluar__ (perhentian keluar), untuk mengidentifikasi nama perhentian
tapOutStopsName|`TapOut_Name`| object | __Nama pintu keluar__ tempat _Customer_ _tap-out_ kartu.
tapOutStopsLat|`TapOut_Lat`| float64 | __Garis Lintang__ _tap-out_ kartu di pintu masuk
tapOutStopsLon|`TapOut_Lon`| float64 | __Garis Bujur__ _tap-out_ kartu di pintu masuk
stopEndSeq|`TapOut_Seq`| float64 | __Urutan pemberhentian__, pemberhentian pertama, pemberhentian kedua, dst. Terkait dengan arah.
tapOutTime|`TapOut_Time`| datetime64[ns] | __Waktu__ ketika _Customer_ keluar (_tap-out_ kartu). __Tanggal dan detail waktu__.
payAmount|`PayAmount`| float64 | __Jumlah uang yang dibayar__ _Customer_. Beberapa gratis. Beberapa tidak.

In [6]:
pd.set_option('display.max_colwidth', 200)

# data unik di tiap kolom
list_kolom = []
for col in df.columns :
    list_kolom.append( [col, df[col].nunique(), df[col].unique()])

TabelDeskripsi = pd.DataFrame(columns=['Column Name', 'Number of Unique', 'Unique Sample'], data=list_kolom)
TabelDeskripsi

Column Name  Number of Unique  \
0        TransID             37900   
1         CardID              2000   
2    PayCardBank                 6   
3           Name              1993   
4         Gender                 2   
5      BirthYear                67   
6     CorridorID               221   
7   CorridorName               216   
8      Direction                 2   
9       TapIn_ID              2570   
10    TapIn_Name              2602   
11     TapIn_Lat              2587   
12     TapIn_Lon              2458   
13     TapIn_Seq                67   
14    TapIn_Time             37079   
15     TapOut_ID              2230   
16   TapOut_Name              2248   
17    TapOut_Lat              2237   
18    TapOut_Lon              2139   
19    TapOut_Seq                74   
20   TapOut_Time             35908   
21     PayAmount                 3   

                                                                                                                                                                                              Unique Sample  
0   [EIIW227B8L34VB, LGXO740D2N47GZ, DJWR385V2U57TO, JTUZ800U7C86EH, VMLO535V7F95NJ, DDES630K2F80KC, HEMW326B9N91TV, XTKE052E5E87LN, OIHS248V7S72EB, ERXO372B2H63RB, CLWK627W2M22WO, IIEK980X1O23NZ, OPL...  
1   [180062659848800, 4885331907664776, 4996225095064169, 639099174703, 570928206772, 2251412124634980, 213155822653833, 3587341902618993, 6510013988638519, 3580401035990896, 5541851157087967, 3472043...  
2                                                                                                                                                                 [emoney, dki, flazz, online, brizzi, bni]  
3   [Bajragin Usada, Gandi Widodo, Emong Wastuti, Surya Wacana, Embuh Mardhiyah, Tirta Siregar, R. Devi Hariyah, S.T., Darmanto Rajasa, Cagak Maheswara, S.Sos, T. Dadap Pradana, M.Ak, Oskar Widiastuti...  
4                                                                                                                                                                                                    [M, F]  
5   [2008, 1997, 1992, 1978, 1982, 1993, 1974, 1991, 2004, 1984, 1988, 2002, 1975, 2011, 2003, 1981, 2009, 1983, 1999, 2007, 1979, 1960, 1995, 2001, 1990, 1968, 1985, 2010, 1989, 2006, 1986, 1965, 200...  
6   [5, 6C, R1A, 11D, 12, 1T, JAK.18, nan, B14, 1Q, 5M, M7B, 9A, 4, JAK.46, S22, 13, 9E, M9, JAK.88, T11, JAK.16, 6V, 11Q, JAK.02, 11, 1C, 3B, M1, 7E, 1P, 7D, 6H, 14, BW9, JIS3, 2, JAK.15, 6N, M11, JA...  
7   [Matraman Baru - Ancol, Stasiun Tebet - Karet via Patra Kuningan, Pantai Maju - Kota, Pulo Gebang - Pulo Gadung 2 via PIK, Tanjung Priok - Pluit, Cibubur - Balai Kota, Kalibata - Kuningan, nan, Be...  
8                                                                                                                                                                                                [1.0, 0.0]  
9   [P00142, B01963P, B00499P, B05587P, P00239, B00127P, B00243P, B03416P, B00795P, B05781P, B05433P, P00254, P00281, P00159, B00608P, B05291P, P00096, B00058P, P00093, B00722P, P00206, nan, P00259, B...  
10  [Pal Putih, Kemenkes 2, Gg. Kunir II, Taman Elok 1, Sunter Boulevard Barat, Benhil 3, Cervino Village, SDN Pondok Labu 11, Imperium, TPU Bungur, Pegangsaan Timur, Tendean, Slipi Petamburan Arah Se...  
11  [-6.184631, -6.2287, -6.133132, -6.195743, -6.14965, -6.21601, -6.224355, -6.313269, -6.210363, -6.247208, -6.199263, -6.240213, -6.201948, -6.207125, -6.284615, -6.319946, -6.236104, -6.232197, -...  
12  [106.84402, 106.83302, 106.81435, 106.93526, 106.889, 106.81632, 106.85165, 106.80221, 106.83059, 106.77889, 106.84205, 106.83102, 106.79998, 106.84112, 106.83787, 106.74598, 106.75488, 106.78874,...  
13  [7, 13, 38, 23, 5, 3, 25, 2, 1, 15, 17, 12, 4, 10, 19, 20, 6, 18, 28, 0, 16, 11, 9, 41, 14, 29, 40, 8, 45, 26, 48, 42, 21, 46, 34, 32, 22, 35, 47, 31, 30, 39, 24, 44, 52, 56, 36, 27, 33, 50, 37, 6...  
14  [2023-04-03 05:21:44, 2023-04-03 05:

Sampel dataset `transjakarta.csv` ini diambil dari bulan April 2024, secara umum di dalamnya berisi 37900 baris dan 21 kolom. Nilai unik dari beberapa kolom cukup banyak (>1000). Kolom yang paling lengkap adalah kolom `TransID` yang mana merupakan tiket customer di setiap transaksi perjalanan menggunakan layanan Transjakarta. Hal ini dianggap wajar karena dari 37900 baris yang ada di dataset, nilai transaksi haruslah berbeda yang menunjukkan setiap baris merupakan perjalanan yang berbeda (unik).

### __C. Cek Data Duplikat__

Selanjutnya kita perlu cek data duplikat untuk memastikan kembali bahwa setiap barisnya unik.

In [7]:
print('Jumlah data duplikat:', df[df.duplicated()].shape[0])

Jumlah data duplikat: 0


> Tidak ada data duplikat

### __D. Total Missing Value Tiap Kolom__

In [8]:
pd.DataFrame(data=df.isna().sum(), columns=['Jumlah Nilai Null'])

Jumlah Nilai Null
TransID                       0
CardID                        0
PayCardBank                   0
Name                          0
Gender                        0
BirthYear                     0
CorridorID                 1257
CorridorName               1930
Direction                     0
TapIn_ID                   1213
TapIn_Name                    0
TapIn_Lat                     0
TapIn_Lon                     0
TapIn_Seq                     0
TapIn_Time                    0
TapOut_ID                  2289
TapOut_Name                1344
TapOut_Lat                 1344
TapOut_Lon                 1344
TapOut_Seq                 1344
TapOut_Time                1344
PayAmount                  1007

### __E. Ringkasan__

Dataset `transjakarta.csv` memiliki jumlah baris sebanyak `37900` buah dan jumlah kolom sebanyak `22` buah. Tipe data dari dataset ini bervariasi dari object, int64, float64, dan datetime64[ns]. _Memory_ yang digunakan dataset ini sebesar 6.4+ MB mengingat jumlah data yang cukup banyak. Jika dilihat dari 5 baris pertamanya secara kasar, tipe data tiap kolomnya sebagian besar merupakan __`kategorikal`__.

Kolom yang memiliki nilai _missing value_:
- `RouteID`
- `RouteName`
- `In_ID`
- `Out_ID`
- `Out_Name`
- `Out_Lat`
- `Out_Lon`
- `Out_Seq`
- `Out_Time`
- `PayAmount`



Karena nilai _missing value_ yang cukup banyak, maka perlu dilakukan observasi terlebih dahulu dan selanjutnya _handling_ pada tiap kolomnya.

---

# ___Data Cleaning___

In [9]:
null = df.isna().sum()
print(null[null>0])

CorridorID      1257
CorridorName    1930
TapIn_ID        1213
TapOut_ID       2289
TapOut_Name     1344
TapOut_Lat      1344
TapOut_Lon      1344
TapOut_Seq      1344
TapOut_Time     1344
PayAmount       1007
dtype: int64


Dari informasi di atas, jumlah nilai null di beberapa kolom cukup banyak. Observasi di setiap kolom diperlukan untuk melihat pola hilangnya data-data tersebut. Pertama kita lihat salah satu contoh nilai null berdasarkan nama _customer_ `Baktiadi Nasyiah`

In [10]:
pd.set_option('display.max_columns', None)
BN = df[df['Name']=='Baktiadi Nasyiah'].sort_values('Direction')
BN

TransID            CardID PayCardBank              Name Gender  \
21521  YZTU719K6E96TF  2265285227934786         dki  Baktiadi Nasyiah      F   
30521  UDVR144Z6N34GJ  2265285227934786         dki  Baktiadi Nasyiah      F   
29521  IDSB018K5V19GN  2265285227934786         dki  Baktiadi Nasyiah      F   
28521  FSSY265I8E51KA  2265285227934786         dki  Baktiadi Nasyiah      F   
27521  BTFE410P9D40DU  2265285227934786         dki  Baktiadi Nasyiah      F   
26521  NKRA582T3P34NO  2265285227934786         dki  Baktiadi Nasyiah      F   
25521  ZBZL454I6B77MN  2265285227934786         dki  Baktiadi Nasyiah      F   
24521  OFRR094N6C90RB  2265285227934786         dki  Baktiadi Nasyiah      F   
23521  TTOT249O1Y19RU  2265285227934786         dki  Baktiadi Nasyiah      F   
22521  JYDE284Z4W72SM  2265285227934786         dki  Baktiadi Nasyiah      F   
20521  EMYP594D1O49HD  2265285227934786         dki  Baktiadi Nasyiah      F   
19521  ZHUP694M4K69XR  2265285227934786         dki  Baktiadi Nasyiah      F   
18521  HWXH633L5R96LP  2265285227934786         dki  Baktiadi Nasyiah      F   
31521  MVXL338P9T17HS  2265285227934786         dki  Baktiadi Nasyiah      F   
17521  DJGI537Q1H41QR  2265285227934786         dki  Baktiadi Nasyiah      F   
12521  NRHL438E9J24UA  2265285227934786         dki  Baktiadi Nasyiah      F   
13521  EIHK215Y5P73TS  2265285227934786         dki  Baktiadi Nasyiah      F   
16521  SXMX730V9V22TU  2265285227934786         dki  Baktiadi Nasyiah      F   
15521  BNEK658U2Q32CV  2265285227934786         dki  Baktiadi Nasyiah      F   
14521  SVBY410A5O87BF  2265285227934786         dki  Baktiadi Nasyiah      F   
19021  KRMZ741J6T49UH  2265285227934786         dki  Baktiadi Nasyiah      F   
30021  TQYK938H5V72GP  2265285227934786         dki  Baktiadi Nasyiah      F   
13021  VTSR669Z8C01KA  2265285227934786         dki  Baktiadi Nasyiah      F   
29021  EGAB572U6T47EU  2265285227934786         dki  Baktiadi Nasyiah      F   
28021  QCCF852S1Z65WC  2265285227934786         dki  Baktiadi Nasyiah      F   
14021  ZTYL380G8X24LJ  2265285227934786         dki  Baktiadi Nasyiah      F   
27021  KBUY877J4W21DV  2265285227934786         dki  Baktiadi Nasyiah      F   
26021  AUZG532P2U53CX  2265285227934786         dki  Baktiadi Nasyiah      F   
25021  CASY691W6N48BW  2265285227934786         dki  Baktiadi Nasyiah      F   
18021  PPCP274F1K82VL  2265285227934786         dki  Baktiadi Nasyiah      F   
24021  HALF684F9Z21EF  2265285227934786         dki  Baktiadi Nasyiah      F   
16021  SLEK232B3R61IQ  2265285227934786         dki  Baktiadi Nasyiah      F   
23021  MNVK048E3F29DQ  2265285227934786         dki  Baktiadi Nasyiah      F   
22021  GJRX421X1W40WE  2265285227934786         dki  Baktiadi Nasyiah      F   
31021  IRZW109P7T10UV  2265285227934786         dki  Baktiadi Nasyiah      F   
21021  PJHO274N0N42ZK  2265285227934786         dki  Baktiadi Nasyiah      F   
17021  JYBB994G5V22NK  2265285227934786         dki  Baktiadi Nasyiah      F   
20021  DSAW776K1Z27TV  2265285227934786         dki  Baktiadi Nasyiah      F   
15021  PRZJ877D9M70JS  2265285227934786         dki  Baktiadi Nasyiah      F   
12021  RLNZ056C6O65CJ  2265285227934786         dki  Baktiadi Nasyiah      F   

       BirthYear CorridorID                         CorridorName  Direction  \
21521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
30521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
29521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
28521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
27521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
26521       1991        NaN  Kalideres - Bundaran HI via Veteran        0.0   
25521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
24521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
23521       1991      

Dari DataFrame di atas, selama bulan April 2023 'Baktiadi Nasyiah' merupakan seorang wanita berumur kurang lebih 32 tahun yang melakukan perjalanan pulang pergi menggunakan layanan transjakarta dengan `CorridorID`: 3 (Rute: Kalideres - Bundaran HI via Veteran). Di pagi hari ia melakukan perjalanan dari __Jelambar__ ke __Petojo__, dan di sore hari ia melakukan perjalanan dari __Dispenda Samsat Barat__ ke __Pesakih__. Biaya yang ia keluarkan tiap perjalanan sebesar __Rp3.500__ dengan pembayaran menggunakan bank __dki__.

Jika diperhatikan terdapat nilai _missing value_ pada kolom `CorridorID`, ataupun kedua `CorridorID` dan `CorridorName`. Ada juga nilai _missing value_ pada kolom `TapIn_ID`, _missing value_ juga ditemukan pada kolom-kolom ketika ia seharusnya melakukan _TapOut_ (untuk mengakhiri perjalanan). Selanjutnya ketika diperhatikan lebih seksama, pada kolom lattitude (`TapIn_Lat`, `TapOut_Lat`) dan longitude (`TapIn_Lon`, `TapOut_Lon`) nilainya selalu sama baik ketika saat _TapIn_ atau _TapOut_ di setiap barisnya, artinya Baktiadi Nasyiah selalu melakukan _TapIn_ dan _TapOut_ di tempat yang sama secara rutin di bulan April.

Nilai _misisng value_ pada kolom `CorridorID` memiliki indikasi terjadi error saat input data. Dengan asumsi jejak histori yang sama, nilai null dapat diisi dengan nilai yang sama dengan historinya (modus), dengan catatan posisi lattitude, longitude, TapIn atau TapOut harus sama. Sebelum mengisi `CorridorID` dan `CorridorName` yang bernilai null, observasi di kolom `TapIn_ID` diperlukan untuk mengecek jejak histori dari _customer_.

---

### __A. Kolom _TapIn_ (`TapIn_ID` & `TapIn_Name`)__ 

In [11]:
print('Jumlah nilai unik pada kolom `TapIn_ID`  :', df['TapIn_ID'].nunique())
print('Jumlah nilai unik pada kolom `TapIn_Name`:', df['TapIn_Name'].nunique())

Jumlah nilai unik pada kolom `TapIn_ID`  : 2570
Jumlah nilai unik pada kolom `TapIn_Name`: 2602


Dari pengecekan sederhana tersebut diketahui bahwa terdapat `TapIn_Name` yang masih belum memiliki/dipasangkan dengan `TapIn_ID`. Pada dasarnya nilai `TapIn_ID` haruslah sepasang dengan nilai `TapIn_Name` karena `TapIn_ID` menjadi kunci untuk nama halte pemberhentian (`TapIn_Name`) layanan Transjakarta. Step selanjutnya dalah mengecek bilamana ada `TapIn_ID` yang berpsangan lebih dari dengan `TapIn_Name` dan sebaliknya.

In [12]:
TapIn_ID_pivot = df.pivot_table(
    index='TapIn_ID',
    values='TapIn_Name',
    aggfunc='nunique'
)
TapInID = TapIn_ID_pivot[TapIn_ID_pivot['TapIn_Name']>1]
TapInID

Empty DataFrame
Columns: [TapIn_Name]
Index: []

> Tidak ada `TapIn_ID` (distinct) yang dipasangkan lebih dari 1 `TapIn_Name`

In [13]:
TapIn_Name_pivot = df.pivot_table(
    index='TapIn_Name',
    values='TapIn_ID',
    aggfunc='nunique'
)
TapInName = TapIn_Name_pivot[TapIn_Name_pivot['TapIn_ID']>1]
TapInName

TapIn_ID
TapIn_Name            
Dukuh Atas 2         2

> Terdapat `TapIn_Name`: 'Dukuh Atas 2' yang memiliki lebih dari 1 `TapIn_ID`

In [14]:
list_TapIn_Name = []
for indeks in TapInName.index:
    list_TapIn_Name.append([indeks, df[df['TapIn_Name']==indeks]['TapIn_ID'].unique()])

pd.DataFrame(data=list_TapIn_Name, columns=['TapIn_Name', 'TapIn_ID'])


TapIn_Name          TapIn_ID
0  Dukuh Atas 2  [P00052, P00298]

Karena terdapat 2 `TapIn_ID` yang dipsangkan dengan 1 `TapIn_Name` yang sama, maka kita perlu observasi lebih lanjut.

In [15]:
df[(df['TapIn_ID']=='P00052') | (df['TapIn_ID']=='P00298')].sort_values('CorridorID', ascending=False)

TransID        CardID PayCardBank                  Name Gender  \
33176  RVXV189M7J06ND  501834837947       flazz      Cahyanto Maulana      M   
13466  MDKU383Q9M74FS  639048791417       flazz  Gantar Namaga, S.Sos      F   
31466  PTCT199E8L29TO  639048791417       flazz  Gantar Namaga, S.Sos      F   
30466  BODS332F8W27OM  639048791417       flazz  Gantar Namaga, S.Sos      F   
29466  XWPI195I5Z32XM  639048791417       flazz  Gantar Namaga, S.Sos      F   
28466  CBEE472M6G46AD  639048791417       flazz  Gantar Namaga, S.Sos      F   
27466  XEZD903N3B39VS  639048791417       flazz  Gantar Namaga, S.Sos      F   
26466  ULUE456X4Z85IC  639048791417       flazz  Gantar Namaga, S.Sos      F   
25466  LUMC322D0I02HM  639048791417       flazz  Gantar Namaga, S.Sos      F   
24466  ZSKO785T6L75MI  639048791417       flazz  Gantar Namaga, S.Sos      F   
12466  YQZR386J1X76PZ  639048791417       flazz  Gantar Namaga, S.Sos      F   
22466  MWKB726R4M78FO  639048791417       flazz  Gantar Namaga, S.Sos      F   
21466  CEHN700I3L62FS  639048791417       flazz  Gantar Namaga, S.Sos      F   
20466  PYOG880C4A77KX  639048791417       flazz  Gantar Namaga, S.Sos      F   
19466  HVKJ261U1U78AG  639048791417       flazz  Gantar Namaga, S.Sos      F   
18466  EHIT887E5A67HI  639048791417       flazz  Gantar Namaga, S.Sos      F   
17466  MHUM533H9N50OV  639048791417       flazz  Gantar Namaga, S.Sos      F   
16466  PTWX505M7Q47XV  639048791417       flazz  Gantar Namaga, S.Sos      F   
15466  GRBH066C3K26TE  639048791417       flazz  Gantar Namaga, S.Sos      F   
23466  IPVY558Q5Q57UA  639048791417       flazz  Gantar Namaga, S.Sos      F   
14466  AQNW655P6W94PI  639048791417       flazz  Gantar Namaga, S.Sos      F   

       BirthYear CorridorID                  CorridorName  Direction TapIn_ID  \
33176       1985         M4  Pulo Gadung 2 - Dukuh Atas 2        1.0   P00298   
13466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
31466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
30466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
29466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
28466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
27466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
26466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
25466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
24466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
12466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
22466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
21466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
20466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
19466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
18466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
17466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
16466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
15466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
23466       1989          4        Pulo Gadung 2 - Tosari        1.0   P00052   
14466       1989        NaN                           NaN        1.0   P00052   

         TapIn_Name  TapIn_Lat   TapIn_Lon  TapIn_Seq          TapIn_Time  \
33176  Dukuh Atas 2  -6.204494  106.823327          0 2023-04-15 06:55:23   
13466  Dukuh Atas 2  -6.203276  106.823770          0 2023-04-04 08:08:41   
31466  Dukuh Atas 2  -6.203276  106.823770          0 2023-04-28 06:47:38   
30466  Dukuh Atas 2  -6.203276  106.823770          0 2023-04-27 09:30:41   
29466  Dukuh Atas 2  -6.203276  106.823770          0 2023-04-26 08:43:58   
28466  D

Ternyata yang membedakan terlihat jelas adalah `CorridorID`/`CorridorName`-nya, dan setelah dicek berdasarkan data aktual transjakarta.co.id, `CorridorID` dengan penulisan `M` terlebih dahulu merupakan layanan __AMARI__ (Angkutan Malam Hari) yang melayani perjalanan malam hari dengan rute yang sama dengan rute utamanya (misalkan: 1 dan M1, 3 dan M3, 5 dan M5, dll.). Normalnya AMARI beroperasi hanya di malam hari, tepatnya pada pukul 22.00 hingga 05.00 keesokan paginya. 

Karena nilainya harus unik dan agar memudahkan proses di tahap analisis, maka `TapIn_Name` dengan `TapIn_ID` P00298 akan di-_rename_ dari 'Dukuh Atas 2' menjadi 'Dukuh Atas 2 (M4)' berdasarkan `CorridorID`-nya.

In [16]:
# Karena hanya 1, maka digunakan metode .loc
df.loc[33176, 'TapIn_Name'] = 'Dukuh Atas 2 (M4)'

In [17]:
# Hasil
df[(df['TapIn_ID']=='P00298')]

TransID        CardID PayCardBank              Name Gender  \
33176  RVXV189M7J06ND  501834837947       flazz  Cahyanto Maulana      M   

       BirthYear CorridorID                  CorridorName  Direction TapIn_ID  \
33176       1985         M4  Pulo Gadung 2 - Dukuh Atas 2        1.0   P00298   

              TapIn_Name  TapIn_Lat   TapIn_Lon  TapIn_Seq  \
33176  Dukuh Atas 2 (M4)  -6.204494  106.823327          0   

               TapIn_Time TapOut_ID    TapOut_Name  TapOut_Lat  TapOut_Lon  \
33176 2023-04-15 06:55:23    P00152  Pasar Genjing   -6.194446   106.86092   

       TapOut_Seq         TapOut_Time  PayAmount  
33176         4.0 2023-04-15 07:53:38     3500.0

Karena nilai `TapIn_ID` dan `TapIn_Name` harusnya sepasang, maka bisa dilakukan pengecekan beberapa kondisi berikut.

In [18]:
# TapIn_ID not null & TapIn_Name null
df[(df['TapIn_ID'].notnull()) & (df['TapIn_Name'].isnull())].shape[0]

0

> Tidak ada di dalam dataset bahwa `TapIn_ID` not null dan `TapIn_Name` null

In [19]:
# TapIn_ID null & TapIn_Name not null
df[(df['TapIn_ID'].isnull()) & (df['TapIn_Name'].notnull())].shape[0]

1213

> Terdapat 1213 baris di dataset yang `TapIn_ID` null dan `TapIn_Name` not null

In [20]:
# TapIn_ID not null & TapIn_Name not null
df[(df['TapIn_ID'].notnull()) & (df['TapIn_Name'].notnull())].shape[0]

36687

> Terdapat 36687 baris dari dataset di mana `TapIn_ID` & `TapIn_Name` not null.

In [21]:
# TapIn_ID null & TapIn_Name null
df[(df['TapIn_ID'].isnull()) & (df['TapIn_Name'].isnull())].shape[0]

0

> Tidak ada di dalam dataset yang `TapIn_ID` dan `TapIn_Name` null

In [22]:
# Membuat Dataframe yang kolom TapIn_ID dan TapIn_Name yang bernilai notnull
df_TapIn = df[(df['TapIn_ID'].notnull()) & (df['TapIn_Name'].notnull())]

# Membuat Dictionary pasangan TapIn_Name(key) dan TapIn_ID(value)
dict_TapIn = {}
for indeks, data in df_TapIn.iterrows():
    dict_TapIn[data['TapIn_Name']] = data['TapIn_ID']

len(dict_TapIn)

2570

> Terdapat 2570 pasangan `TapIn_ID` dan `TapIn_Name`

In [23]:
print('Jumlah nilai unik pada kolom `TapIn_ID`     :', df['TapIn_ID'].nunique())
print('Jumlah nilai unik pada kolom `TapIn_Name`   :', df['TapIn_Name'].nunique())
print('Jumlah missing value pada kolom `TapIn_ID`  :', df['TapIn_ID'].isna().sum())
print('Jumlah missing value pada kolom `TapIn_Name`:', df['TapIn_Name'].isna().sum())

Jumlah nilai unik pada kolom `TapIn_ID`     : 2570
Jumlah nilai unik pada kolom `TapIn_Name`   : 2603
Jumlah missing value pada kolom `TapIn_ID`  : 1213
Jumlah missing value pada kolom `TapIn_Name`: 0


Karena nilainya berbeda, maka perlu dicek `TapIn_Name` mana saja yang tidak memiliki `TapIn_ID`

In [24]:
TapIn_Name_unique = list(df['TapIn_Name'].unique())

In [25]:
# membuat set `TapIn_Name` yang tidak memiliki `TapIn_ID`
no_TapIn_ID = set()
for name in TapIn_Name_unique:
    if name not in dict_TapIn:
        no_TapIn_ID.add(name)

len(no_TapIn_ID)

33

> Terdapat 33 `TapIn_Name` yang belum memiliki ID

Karena nilai `TapIn_ID` adalah unik, kita bisa _generate_ `TapIn_ID` baru untuk dipasangkan dengan `TapIn_Name` yang belum memiliki `TapIn_ID` dengan awalan `N` sebagai pembeda.

In [26]:
# Generate TapIn_ID baru
new_In_ID = []
for i in range(1, 34):
    new_In_ID.append(f"N{i:05}")

# Memasangkan TapIn_ID baru di-generate dengan TapIn_Name yang belum memiliki ID
dict_TapIn_new = {}
for x, y in zip(list(no_TapIn_ID), new_In_ID):
    dict_TapIn_new[x] = y

Selanjutnya isi nilai baru tersebut ke DataFrame induk (df).

In [27]:
for indeks, data in df.iterrows():
    if data['TapIn_Name'] in dict_TapIn_new.keys():
        df.loc[indeks, 'TapIn_ID'] = dict_TapIn_new[data['TapIn_Name']]

In [28]:
print('Jumlah nilai unik pada kolom `TapIn_ID`     :', df['TapIn_ID'].nunique())
print('Jumlah nilai unik pada kolom `TapIn_Name`   :', df['TapIn_Name'].nunique())
print('Jumlah missing value pada kolom `TapIn_ID`  :', df['TapIn_ID'].isna().sum())
print('Jumlah missing value pada kolom `TapIn_Name`:', df['TapIn_Name'].isna().sum())

Jumlah nilai unik pada kolom `TapIn_ID`     : 2603
Jumlah nilai unik pada kolom `TapIn_Name`   : 2603
Jumlah missing value pada kolom `TapIn_ID`  : 1179
Jumlah missing value pada kolom `TapIn_Name`: 0


Masih terdapat __missing value__ pada kolom `TapIn_ID` karena belum diisi berdasarkan `TapIn_Name`

In [29]:
# Mengisi nilai kosong di kolom TapIn_ID berdasarkan TapIn_Name
for indeks, data in df.iterrows():
    if pd.isnull(data['TapIn_ID']):
        df.loc[indeks, 'TapIn_ID'] = dict_TapIn[data['TapIn_Name']]

print('Jumlah missing value pada kolom `TapIn_ID`  :', df['TapIn_ID'].isna().sum())
print('Jumlah missing value pada kolom `TapIn_Name`:', df['TapIn_Name'].isna().sum())

Jumlah missing value pada kolom `TapIn_ID`  : 0
Jumlah missing value pada kolom `TapIn_Name`: 0


> Sudah tidak ada missing value pada kolom `TapIn_ID` dan `TapIn_Name`

---

### __B. Kolom Rute (`CorridorID` & `CorridorName`)__ 

Sama halnya dengan kolom _TapIn_ nilai `CorridorID` haruslah sepasang dengan nilai `CorridorName` karena `CorridorID` menjadi kunci untuk pengelompokkan rute perjalanan. Namun dari dataset `transjakarta.csv`, perlu dicek apakah hal ini benar.

In [30]:
CorridorID_pivot = df.pivot_table(
    index='CorridorID',
    values='CorridorName',
    aggfunc='nunique'
)
CorridorID_pivot[CorridorID_pivot['CorridorName']>1]

Empty DataFrame
Columns: [CorridorName]
Index: []

> Dari dataset tidak ada `CorridorID` yang memiliki lebih dari 1 `CorridorName`

Dengan begitu kita bisa mengisi kolom `Corridor` yang bernilai null bila terdapat nilai (not null) pada kolom `CorridorID`

In [31]:
# List seluruh CorridorID dari DataFrame
list_CorridorID = []
for id in df['CorridorID']:
    if pd.isna(id) or id in list_CorridorID:
        continue
    else:
        list_CorridorID.append(str(id))

print('Jumlah CorridorID yang ada di dataset:', len(list_CorridorID))

Jumlah CorridorID yang ada di dataset: 221


In [32]:
# Membuat dataframe dengan kondisi pasangan kolom CorridorID dan CorridorName yang bukan missing value
df_corridor = df[(df['CorridorID'].notnull()) & (df['CorridorName'].notnull())]

In [33]:
# Membuat Dictionary pasangan CorridorID(key) dan CorridorName(value)
dict_CorridorID = {}
for indeks, data in df_corridor.iterrows():
    dict_CorridorID[data['CorridorID']] = data['CorridorName']

print('Jumlah pasangan CorridorID dan CorridorName yang ada di dataset:', len(dict_CorridorID))
print('Jumlah CorridorID  :', df['CorridorID'].nunique())
print('Jumlah CorridorName:', df['CorridorName'].nunique())

Jumlah pasangan CorridorID dan CorridorName yang ada di dataset: 221
Jumlah CorridorID  : 221
Jumlah CorridorName: 216


In [34]:
# Mengisi kolom CorridorName yang bernilai null berdasarkan CorridorID
for indeks, data in df.iterrows():
    if pd.notnull(data['CorridorID']) and pd.isnull(data['CorridorName']):
        df.loc[indeks, 'CorridorName'] = dict_CorridorID[data['CorridorID']]

In [35]:
# Pengecekan kembali
df[(df['CorridorID'].notnull()) & (df['CorridorName'].isnull())].shape[0]

0

Berikutnya dicek jumlah CorridorName yang unik berdasarkan CorridorID

In [36]:
CorridorName_pivot = df.pivot_table(
    index='CorridorName',
    values='CorridorID',
    aggfunc='nunique'
)
CorridorID_2 = CorridorName_pivot[CorridorName_pivot['CorridorID']>1]
CorridorID_2.shape[0]
# Terdapat Corridor yang memiliki lebih dari 1 RouteID

5

> Terdapat 5 `CorridorName` yang memiliki pasangan lebih dari 1 dengan `CorridorID`

In [37]:
# Cek Nilai CorridorID-nya
list_CorridorName = []
for indeks in CorridorID_2.index:
    list_CorridorName.append([indeks, df[df['CorridorName']==indeks]['CorridorID'].unique()])

pd.DataFrame(data=list_CorridorName, columns=['CorridorName', 'CorridorID'])

CorridorName    CorridorID
0                        Blok M - Kota       [M1, 1]
1  Kalideres - Bundaran HI via Veteran  [M3, 3, nan]
2                Matraman Baru - Ancol  [5, nan, M5]
3                 Pinang Ranti - Pluit  [M9, 9, nan]
4                  Pulo Gadung - Monas  [2, M2, nan]

Secara garis besar nilai `CorridorName` yang memiliki lebih dari 1 `CorridorID` merupakan layana AMARI dari transjakarta yang rutenya sama dengan rute utama (misalkan: 1 dengan M1, 2 dengan M2, 3 dengan M3). Karena kasusnya sama dengan sebelumnya, agar menjadi pembeda dan memudahkan proses analisis dataset, maka akan dilakukan _rename_ CorridorName layanan AMARI dengan CorridorID-nya. Contoh: `Blok M - Kota` menjadi `Blok M - Kota (M1)`

In [38]:
list_CorridorID_AMARI = ['M1', 'M2', 'M3', 'M5', 'M9']
list_CorridorName_AMARI = ['Blok M - Kota (M1)', 'Pulo Gadung - Monas (M2)', 'Kalideres - Bundaran HI via Veteran (M3)', 'Matraman Baru - Ancol (M5)', 'Pinang Ranti - Pluit (M9)']

for id, name in zip(list_CorridorID_AMARI, list_CorridorName_AMARI):
    indeks = df[df['CorridorID']==id].index
    for idx in indeks:
        df.loc[idx, 'CorridorName'] = name

In [39]:
print('Jumlah CorridorID  :', df['CorridorID'].nunique())
print('Jumlah CorridorName:', df['CorridorName'].nunique())
print('Jumlah missing value pada kolom `CorridorID`  :', df['CorridorID'].isna().sum())
print('Jumlah missing value pada kolom `CorridorName`:', df['CorridorName'].isna().sum())

Jumlah CorridorID  : 221
Jumlah CorridorName: 221
Jumlah missing value pada kolom `CorridorID`  : 1257
Jumlah missing value pada kolom `CorridorName`: 1125


In [40]:
df[(df['CorridorID']=='M2')].index

Index([   92,   392,   692,   992,  1292,  1592,  1892,  2192,  2492,  2792,
        3092,  3392,  3692,  3992,  4292,  4592,  4892,  5192,  5492,  5792,
        6092,  6392,  6692,  6992,  7292,  7592,  7892,  8192,  8492,  8792,
        9092,  9392,  9692,  9992, 10292, 10892, 11192, 11492, 11792, 32387,
       32917, 32992, 33226, 33344, 33399, 33454, 34175, 34874, 35074, 35274,
       35474, 35674, 35874, 36074, 36274, 36474, 36674, 36874, 37074, 37274,
       37474],
      dtype='int64')

In [41]:
df[(df['CorridorID']=='M2')&(df['CorridorName']=='Pulo Gadung - Monas')]

Empty DataFrame
Columns: [TransID, CardID, PayCardBank, Name, Gender, BirthYear, CorridorID, CorridorName, Direction, TapIn_ID, TapIn_Name, TapIn_Lat, TapIn_Lon, TapIn_Seq, TapIn_Time, TapOut_ID, TapOut_Name, TapOut_Lat, TapOut_Lon, TapOut_Seq, TapOut_Time, PayAmount]
Index: []

> Sudah tidak ada `CorridorName` yang memiliki pasangan lebih dari 1 dengan `CorridorID`

Karena nilai uniknya sudah sama, artinya mereka sudah berpasangan sehingga kita bisa mengisi kolom `CorridorID` yang bernilai null berdasarkan `CorridorName`

In [42]:
# Membuat kondisi pasangan kolom CorridorID dan CorridorName yang bukan missing value
CorridorName_by_ID = df[(df['CorridorID'].notnull()) & (df['CorridorName'].notnull())]

# Membuat Dictionary pasangan CorridorID(key) dan Corridor(value)
dict_CorridorName = {}
for indeks, data in CorridorName_by_ID.iterrows():
    dict_CorridorName[data['CorridorName']] = data['CorridorID']

# Mengisi kolom CorridorID yang bernilai Null berdasarkan CorridorName
for indeks, data in df.iterrows():
    if pd.notnull(data['CorridorName']) and pd.isnull(data['CorridorID']):
        df.loc[indeks, 'CorridorID'] = dict_CorridorName[data['CorridorName']]

print('Jumlah missing value pada kolom `CorridorID`  :', df['CorridorID'].isna().sum())
print('Jumlah missing value pada kolom `CorridorName`:', df['CorridorName'].isna().sum())

Jumlah missing value pada kolom `CorridorID`  : 1125
Jumlah missing value pada kolom `CorridorName`: 1125


In [43]:
# Kondisi CorridorID dan Corridor Name yang bernilai null
df[(df['CorridorID'].isnull())&(df['CorridorName'].isnull())].head()

TransID               CardID PayCardBank                   Name  \
7    XTKE052E5E87LN     3587341902618993         dki        Darmanto Rajasa   
120  MCWO805A6Q66MX     3570416613897992         dki   Gawati Damanik, S.E.   
240  PTDK452M9A71XF      342370004909630      emoney  Kartika Habibi, S.Psi   
250  OLZA217B5B95DK  4371940047525876160      brizzi          Zelda Laksita   
257  GHEY769K9T55WF        4281653494791      online           Zelda Wibowo   

    Gender  BirthYear CorridorID CorridorName  Direction TapIn_ID  \
7        F       1991        NaN          NaN        1.0  B03416P   
120      F       2010        NaN          NaN        0.0   P00109   
240      M       1975        NaN          NaN        0.0  B04505P   
250      M       2009        NaN          NaN        1.0  B04576P   
257      M       1989        NaN          NaN        0.0  B00048P   

                  TapIn_Name  TapIn_Lat  TapIn_Lon  TapIn_Seq  \
7         SDN Pondok Labu 11  -6.313269  106.80221          2   
120            Matraman Baru  -6.212880  106.86115          8   
240  Sbr. Standard Chartered  -6.217407  106.81934         13   
250   Sbr. Univ. Mercu Buana  -6.210273  106.73786         11   
257   Aneka PIK Penggilingan  -6.202636  106.92867         16   

             TapIn_Time TapOut_ID            TapOut_Name  TapOut_Lat  \
7   2023-04-03 06:44:24   B00899P             Jln. Baros   -6.311046   
120 2023-04-03 06:38:44    P00224          Slamet Riyadi   -6.208533   
240 2023-04-03 05:13:16   B03396P        Kota Kasablanka   -6.224688   
250 2023-04-03 06:24:37   B05009P  Simpang Meruya Ilir 1   -6.196682   
257 2023-04-03 05:05:24   B05823P      United Tractors 1   -6.183260   

     TapOut_Lon  TapOut_Seq         TapOut_Time  PayAmount  
7     106.78400        12.0 2023-04-03 07:11:43        0.0  
120   106.85929         9.0 2023-04-03 07:32:11     3500.0  
240   106.84246        20.0 2023-04-03 05:39:55     3500.0  
250   106.75517        28.0 2023-04-03 07:13:18        0.0  
257   106.93243        28.0 2023-04-03 05:41:12     3500.0

Permasalahan selanjutnya ialah dari ke-5 `CorridorName` yang sebelumnya dibahas terdapat nilai null pada kolom `CorridorID` yang tidak bisa kita isi serta merta berdasarkan `CorridorName`-nya, karena 1 halte pemberhentian dapat dilalui oleh beberapa Corridor yang artinya bisa 

In [44]:
pd.set_option('display.max_columns', None)

df[(df['CorridorName']=='Kalideres - Bundaran HI via Veteran') & (df['CorridorID'].isnull())|\
   (df['CorridorName']=='Matraman Baru - Ancol') & (df['CorridorID'].isnull())|\
   (df['CorridorName']=='Pinang Ranti - Pluit') & (df['CorridorID'].isnull())|\
   (df['CorridorName']=='Pulo Gadung - Monas') & (df['CorridorID'].isnull())]

Empty DataFrame
Columns: [TransID, CardID, PayCardBank, Name, Gender, BirthYear, CorridorID, CorridorName, Direction, TapIn_ID, TapIn_Name, TapIn_Lat, TapIn_Lon, TapIn_Seq, TapIn_Time, TapOut_ID, TapOut_Name, TapOut_Lat, TapOut_Lon, TapOut_Seq, TapOut_Time, PayAmount]
Index: []

Kita ambil contoh pada `CorridorName` yang bernilai 'Kalideres - Bundaran HI via Veteran'

In [45]:
df[(df['CorridorName']=='Kalideres - Bundaran HI via Veteran') & (df['CorridorID'].isna())]

Empty DataFrame
Columns: [TransID, CardID, PayCardBank, Name, Gender, BirthYear, CorridorID, CorridorName, Direction, TapIn_ID, TapIn_Name, TapIn_Lat, TapIn_Lon, TapIn_Seq, TapIn_Time, TapOut_ID, TapOut_Name, TapOut_Lat, TapOut_Lon, TapOut_Seq, TapOut_Time, PayAmount]
Index: []

Ternyata orang yang sama 'Baktiadi Nasyiah'. Cek kembali historinya

In [46]:
df[(df['Name']=='Baktiadi Nasyiah')].sort_values('Direction').head(6)

TransID            CardID PayCardBank              Name Gender  \
21521  YZTU719K6E96TF  2265285227934786         dki  Baktiadi Nasyiah      F   
30521  UDVR144Z6N34GJ  2265285227934786         dki  Baktiadi Nasyiah      F   
29521  IDSB018K5V19GN  2265285227934786         dki  Baktiadi Nasyiah      F   
28521  FSSY265I8E51KA  2265285227934786         dki  Baktiadi Nasyiah      F   
27521  BTFE410P9D40DU  2265285227934786         dki  Baktiadi Nasyiah      F   
26521  NKRA582T3P34NO  2265285227934786         dki  Baktiadi Nasyiah      F   

       BirthYear CorridorID                         CorridorName  Direction  \
21521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
30521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
29521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
28521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
27521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   
26521       1991          3  Kalideres - Bundaran HI via Veteran        0.0   

      TapIn_ID TapIn_Name  TapIn_Lat  TapIn_Lon  TapIn_Seq  \
21521   P00087   Jelambar  -6.166548  106.78651          9   
30521   P00087   Jelambar  -6.166548  106.78651          9   
29521   P00087   Jelambar  -6.166548  106.78651          9   
28521   P00087   Jelambar  -6.166548  106.78651          9   
27521   P00087   Jelambar  -6.166548  106.78651          9   
26521   P00087   Jelambar  -6.166548  106.78651          9   

               TapIn_Time TapOut_ID TapOut_Name  TapOut_Lat  TapOut_Lon  \
21521 2023-04-14 17:10:16    P00176      Petojo   -6.169984   106.81698   
30521 2023-04-27 18:39:42    P00176      Petojo   -6.169984   106.81698   
29521 2023-04-26 21:43:26    P00176      Petojo   -6.169984   106.81698   
28521 2023-04-25 17:48:30    P00176      Petojo   -6.169984   106.81698   
27521 2023-04-24 19:27:47    P00176      Petojo   -6.169984   106.81698   
26521 2023-04-21 19:20:37    P00176      Petojo   -6.169984   106.81698   

       TapOut_Seq         TapOut_Time  PayAmount  
21521        12.0 2023-04-14 19:00:06     3500.0  
30521        12.0 2023-04-27 20:23:21     3500.0  
29521        12.0 2023-04-26 23:28:20     3500.0  
28521        12.0 2023-04-25 19:01:03     3500.0  
27521        12.0 2023-04-24 21:06:57     3500.0  
26521        12.0 2023-04-21 20:28:49     3500.0

Sama dengan pembahasan sebelumnya _misisng value_ pada kolom `CorridorID` memiliki indikasi terjadi error saat input data. Dengan asumsi jejak histori yang sama, nilai null dapat diisi dengan nilai yang sama dengan historinya (modus), dengan catatan posisi lattitude, longitude, TapIn atau TapOut harus sama.

In [47]:
# Membagi Customer berdasarkan Direction
list_Name_0 = list(df[df['Direction']==0.0]['Name'].unique())
list_Name_1 = list(df[df['Direction']==1.0]['Name'].unique())

In [48]:
# Mengisi nilai null pada kolom CorridorID berdasarkan jejak histori customer pada Direction (0.0)
for nama in list_Name_0:
    dataframe = df[(df['Name']==nama) & (df['Direction']==0.0)]
    if dataframe.shape[0]>2:
        modus_TapIn_ID_0 = dataframe['TapIn_ID'].mode()[0]
        modus_TapIn_Lat_0 = dataframe['TapIn_Lat'].mode()[0]
        modus_TapIn_Lon_0 = dataframe['TapIn_Lon'].mode()[0]
        modus_TapIn_Seq_0 = dataframe['TapIn_Seq'].mode()[0]

    for indeks, data in dataframe.iterrows():
        if pd.isnull(data['CorridorID']):
            if data['TapIn_ID'] == modus_TapIn_ID_0:
                if data['TapIn_Lat'] == modus_TapIn_Lat_0:
                    if data['TapIn_Lon'] == modus_TapIn_Lon_0:
                        if data['TapIn_Seq'] == modus_TapIn_Seq_0:
                            df.loc[indeks, 'CorridorID'] = dataframe['CorridorID'].mode()[0]

In [49]:
# Mengisi nilai null pada kolom CorridorID berdasarkan jejak histori customer pada Direction (1.0)
for nama in list_Name_1:
    dataframe = df[(df['Name']==nama) & (df['Direction']==1.0)]
    if dataframe.shape[0]>2:
        modus_TapIn_ID_1 = dataframe['TapIn_ID'].mode()[0]
        modus_TapIn_Lat_1 = dataframe['TapIn_Lat'].mode()[0]
        modus_TapIn_Lon_1 = dataframe['TapIn_Lon'].mode()[0]
        modus_TapIn_Seq_1 = dataframe['TapIn_Seq'].mode()[0]

    for indeks, data in dataframe.iterrows():
        if pd.isnull(data['CorridorID']):
            if data['TapIn_ID'] == modus_TapIn_ID_1:
                if data['TapIn_Lat'] == modus_TapIn_Lat_1:
                    if data['TapIn_Lon'] == modus_TapIn_Lon_1:
                        if data['TapIn_Seq'] == modus_TapIn_Seq_1:
                            df.loc[indeks, 'CorridorID'] = dataframe['CorridorID'].mode()[0]

In [50]:
print('Jumlah nilai unik pada kolom `CorridorID`     :', df['CorridorID'].nunique())
print('Jumlah nilai unik pada kolom `CorridorName`   :', df['CorridorName'].nunique())
print('Jumlah missing value pada kolom `CorridorID`  :', df['CorridorID'].isna().sum())
print('Jumlah missing value pada kolom `CorridorName`:', df['CorridorName'].isna().sum())

Jumlah nilai unik pada kolom `CorridorID`     : 221
Jumlah nilai unik pada kolom `CorridorName`   : 221
Jumlah missing value pada kolom `CorridorID`  : 88
Jumlah missing value pada kolom `CorridorName`: 1125


Karena jumlah missing value yang berbeda, digunakan kkode yang sebelumnya untuk mengisi `CorridorName` yang masih bernilai kosong berdasarkan `CorridorID` yang telah diisi sebelumnya

In [51]:
# Mengisi kolom CorridorName yang bernilai null berdasarkan CorridorID
for indeks, data in df.iterrows():
    if pd.notnull(data['CorridorID']) and pd.isnull(data['CorridorName']):
        df.loc[indeks, 'CorridorName'] = dict_CorridorID[data['CorridorID']]

In [52]:
print('Jumlah nilai unik pada kolom `CorridorID`     :', df['CorridorID'].nunique())
print('Jumlah nilai unik pada kolom `CorridorName`   :', df['CorridorName'].nunique())
print('Jumlah missing value pada kolom `CorridorID`  :', df['CorridorID'].isna().sum())
print('Jumlah missing value pada kolom `CorridorName`:', df['CorridorName'].isna().sum())

Jumlah nilai unik pada kolom `CorridorID`     : 221
Jumlah nilai unik pada kolom `CorridorName`   : 221
Jumlah missing value pada kolom `CorridorID`  : 88
Jumlah missing value pada kolom `CorridorName`: 88


88 orang tersebut merupakan orang-orang yang hanya menggunakan minimal 1 kali layanan Transjakarta. Kita tidak bisa mengasumsikan `CorridorID` dan `CorridorName` mana yang ia gunakan. Karena data yang kita punya berasal dari tahun 2023, kita tidak bisa mengecek kembali ke transjakarta.co.id yang sudah melakukan pembaruan data (terdapat rute yang sudah tidak digunakan, berganti nama, berganti jadwal atau posisi pemberhentian). Karena tidak tahu dengan pasti rute mana yang digunakan, dari ke-88 orang (88 baris), akan dihapus.

---

### __C. Kolom _TapOut_ (`TapOut_ID`, `TapOut_Name`, `TapOut_Lat`, `TapOut_Lon`, `TapOut_Seq`)__ 

In [53]:
df[(df['TapOut_ID'].isna()) | (df['TapOut_Name'].isna()) | (df['TapOut_Lat'].isna()) | (df['TapOut_Lon'].isna()) | (df['TapOut_Seq'].isna())].head()

TransID            CardID PayCardBank                       Name  \
6    HEMW326B9N91TV   213155822653833      emoney      R. Devi Hariyah, S.T.   
30   TCKQ601Q5O01BC  3589380857722689         dki      drg. Lukman Hidayanto   
70   OTGH855B8A29BP  3549158005951908         dki  Mustofa Wijayanti, M.Farm   
89   QMHC825R0K38KM  3537543297101185         dki             Samsul Ramadan   
130  JUKK703H7F00XH      570918242399       flazz            Mumpuni Gunarto   

    Gender  BirthYear CorridorID                      CorridorName  Direction  \
6        M       1974     JAK.18               Kalibata - Kuningan        0.0   
30       M       1995     JAK.46          Pasar Minggu - Jagakarsa        1.0   
70       M       1981         1T              Cibubur - Balai Kota        0.0   
89       M       1966        M11       Pulo Gebang - Pulo Gadung 2        1.0   
130      F       1998        10D  Kampung Rambutan - Tanjung Priok        0.0   

    TapIn_ID         TapIn_Name  TapIn_Lat  TapIn_Lon  TapIn_Seq  \
6    B00243P    Cervino Village  -6.224355  106.85165         25   
30   B00001P     18 Office Park  -6.299146  106.83210         19   
70   B02408P       Menara Astra  -6.207361  106.82190          7   
89    P00158  Pasar Pulo Gadung  -6.187434  106.90591          0   
130   P00035        Cempaka Mas  -6.165888  106.87920          7   

             TapIn_Time TapOut_ID             TapOut_Name  TapOut_Lat  \
6   2023-04-03 06:58:21       NaN                     NaN         NaN   
30  2023-04-03 06:14:02       NaN                     NaN         NaN   
70  2023-04-03 05:57:55       NaN                     NaN         NaN   
89  2023-04-03 06:23:36       NaN  Walikota Jakarta Timur    -6.21254   
130 2023-04-03 05:40:04       NaN                     NaN         NaN   

     TapOut_Lon  TapOut_Seq         TapOut_Time  PayAmount  
6           NaN         NaN                 NaT        0.0  
30          NaN         NaN                 NaT        0.0  
70          NaN         NaN                 NaT    20000.0  
89    106.94537         7.0 2023-04-03 06:52:14     3500.0  
130         NaN         NaN                 NaT     3500.0

Yang menjadi perhatian di sini adalah teradapat banyak _missing value_ pada 1 baris. Hal ini bisa terjadi bila user lupa melakukan _TapOut_ ketika hendak keluar, atau mesin untuk melakukan _tapping_ sedang tidak berfungsi dengan benar, alhasil banyak data-data customer saat hendak _TapOut_ tidak ada. Kita bisa menggunakan metode pada kolom-kolom TapIn ('In_ID', 'In_Name', 'In_Lat', 'In_Lon') untuk mengisi nilai yang kosong pada kolom-kolom TapOut. 

Pertama kita cek terlebih dahulu jumlah nilai unik dari kolom `TapOut_ID` dan `TapOut_Name`

In [54]:
print('Jumlah nilai unik di kolom `TapOut_ID`  :', df['TapOut_ID'].nunique())
print('Jumlah nilai unik di kolom `TapOut_Name`:', df['TapOut_Name'].nunique())

Jumlah nilai unik di kolom `TapOut_ID`  : 2230
Jumlah nilai unik di kolom `TapOut_Name`: 2248


Terdapat perbedaan sebanyak 18 buah, artinya tidak semua nilai pada kolom `TapOut_ID` terpasangkan dengan `TapOut_Name`, observasi lebih lanjut diperlukan untuk dilihat polanya.

In [55]:
TapOut_ID_pivot = df.pivot_table(
    index='TapOut_ID',
    values='TapOut_Name',
    aggfunc='nunique'
)
TapOut_ID_pivot[TapOut_ID_pivot['TapOut_Name']>1].shape[0]

0

> Setiap `TapOut_ID` yang ada telah terpasangkan dengan `TapOut_Name`-nya

In [56]:
TapOut_Name_pivot = df.pivot_table(
    index='TapOut_Name',
    values='TapOut_ID',
    aggfunc='nunique'
)
TapOut_ID_2 = TapOut_Name_pivot[TapOut_Name_pivot['TapOut_ID']>1]
TapOut_Name_pivot[TapOut_Name_pivot['TapOut_ID']>1].shape[0]

1

> Terdapat 1 `TapOut_Name` yang memiliki lebih dari 1 `TapOut_ID`

In [57]:
# Cek Nilai TapOut_ID-nya
list_TapOut = []
for indeks in TapOut_ID_2.index:
    list_TapOut.append([indeks, df[df['TapOut_Name']==indeks]['TapOut_ID'].unique()])

double = pd.DataFrame(data=list_TapOut, columns=['TapOut_Name', 'TapOut_ID'])
double

TapOut_Name              TapOut_ID
0  Dukuh Atas 2  [P00052, nan, P00298]

In [58]:
DA2 = df[df['TapOut_Name']=='Dukuh Atas 2']
DA2

TransID            CardID PayCardBank                   Name  \
176    KAGY982E0G63WU  6542837703080439         dki         Aurora Nasyiah   
776    VYCP933N4D90DE  6542837703080439         dki         Aurora Nasyiah   
1376   GYUF829J4P49DH  6542837703080439         dki         Aurora Nasyiah   
1976   WNJZ321W6X93GD  6542837703080439         dki         Aurora Nasyiah   
2576   GJEE297W6K37EZ  6542837703080439         dki         Aurora Nasyiah   
3176   PHGI772L6Q50TU  6542837703080439         dki         Aurora Nasyiah   
3776   YUOT647A2Z19XF  6542837703080439         dki         Aurora Nasyiah   
4376   CMDK510J9B76KG  6542837703080439         dki         Aurora Nasyiah   
4976   TSML556X5W26WJ  6542837703080439         dki         Aurora Nasyiah   
5576   VPXC161P5W80JK  6542837703080439         dki         Aurora Nasyiah   
6176   QGQU213Y8Z38IQ  6542837703080439         dki         Aurora Nasyiah   
6776   OBXF299U1R72RN  6542837703080439         dki         Aurora Nasyiah   
7376   SACS354U7D08UZ  6542837703080439         dki         Aurora Nasyiah   
7976   WZXN484Z2Z54JM  6542837703080439         dki         Aurora Nasyiah   
8576   JYRO877H0N72XP  6542837703080439         dki         Aurora Nasyiah   
9176   FGQI065C4D82IC  6542837703080439         dki         Aurora Nasyiah   
9776   VICF353H5C65GC  6542837703080439         dki         Aurora Nasyiah   
10376  KHYR883B2W33KX  6542837703080439         dki         Aurora Nasyiah   
10976  AXKP383K8J30OC  6542837703080439         dki         Aurora Nasyiah   
11576  YANF054V1I78PG  6542837703080439         dki         Aurora Nasyiah   
12861  KNRQ622P2C20PR  2720497305962063         dki         Adiarja Waluyo   
13861  UYPX200C6Y31RS  2720497305962063         dki         Adiarja Waluyo   
14861  PRXB564T3T11MG  2720497305962063         dki         Adiarja Waluyo   
15861  PTSD360C6E26QP  2720497305962063         dki         Adiarja Waluyo   
16861  NVMZ545H5H82NK  2720497305962063         dki         Adiarja Waluyo   
17861  EOJJ370S5Y31FZ  2720497305962063         dki         Adiarja Waluyo   
18861  ZMAA142H1F66KE  2720497305962063         dki         Adiarja Waluyo   
19861  ABCC851P3R17SM  2720497305962063         dki         Adiarja Waluyo   
20861  ZDOR622U1I23IO  2720497305962063         dki         Adiarja Waluyo   
21861  GROG623K8K28OL  2720497305962063         dki         Adiarja Waluyo   
22861  GKXZ543Q3D23CV  2720497305962063         dki         Adiarja Waluyo   
23861  XENX011S3A86EI  2720497305962063         dki         Adiarja Waluyo   
24861  RKDE923B1F21FE  2720497305962063         dki         Adiarja Waluyo   
25861  CRXP985N2N66GN  2720497305962063         dki         Adiarja Waluyo   
26861  VLHA878Q3X90NA  2720497305962063         dki         Adiarja Waluyo   
27861  HTEA741W6U92KL  2720497305962063         dki         Adiarja Waluyo   
28861  JQNF111L8V38US  2720497305962063         dki         Adiarja Waluyo   
29861  CZGX447Q0E77AV  2720497305962063         dki         Adiarja Waluyo   
30861  VPRB187C1U58FQ  2720497305962063         dki         Adiarja Waluyo   
31861  PFPK305H3W00OX  2720497305962063         dki         Adiarja Waluyo   
32303  ZCQT702V1D66HG  3544914025938688         dki     Cakrabuana Yolanda   
33623  AYLH004T9I37FG      676246072349       flazz  Raharja Hastuti, M.Pd   
34209  TBRJ942F0U06OF      676190181542       flazz       Tgk. Umay Usamah   
34678  OHIA748R9V48ZV    30023091550042         bni          Karsa Hutapea   
35016  JSRN267K0U89MJ   370095866460258      emoney     Pranata Tampubolon   
35416  UBNP698W8I64JU   370095866460258      emoney     Pranata Tampubolon   
35816  HYUP429Y7N07AV   370095866460258      emoney     Pranata Tampubolon   
36216  LKJT085D1N80DJ   370095866460258      emoney     Pranata Tampubolon   
36616  NSCO625G8E21QB   370095866460258      emoney     Pranata Tampubolon   
37016  AYQH654A7J33WC   370095866460258      emoney     Pranata Tampubolon   
37416  KUSP940N4J16XL   370095866460

Mirip dengan kondisi sebelumnya, _rename_ nama halte pemberhentian dari `Dukuh Atas 2` menjadi `Dukuh Atas 2 (6)` pada `Out_ID` P00052 agar mudah dibedakan.

In [59]:
# Rename 'Dukuh Atas 2' menjadi 'Dukuh Atas 2 (M4)'
for indeks in DA2[(DA2['TapOut_ID']=='P00052')&(DA2['CorridorID']=='M4')].index:
    df.loc[indeks, 'TapOut_Name'] = 'Dukuh Atas 2 (M4)'

Cek `TapOut_Name` mana yang belum memiliki `Out_ID`

In [60]:
# Membuat Dataframe yang kolom TapOut_ID dan TapOut_Name yang bernilai notnull
df_TapOut_Name_ID = df[(df['TapOut_ID'].notnull()) & (df['TapOut_Name'].notnull())]

In [61]:
# Membuat Dictionary pasangan TapOut_Name(key) dan TapOut_ID(value)
dict_TapOut_ID_Name = {}
for indeks, data in df_TapOut_Name_ID.iterrows():
    dict_TapOut_ID_Name[data['TapOut_Name']] = data['TapOut_ID']

In [62]:
TapOut_Name_unique = list(df['TapOut_Name'].unique())

In [63]:
no_TapOut_ID = set()
for name in TapOut_Name_unique:
    if name not in dict_TapOut_ID_Name and pd.notnull(name):
        no_TapOut_ID.add(name)
print('Jumlah `TapOut_Name` yang belum memiliki `Out_ID`:', len(no_TapOut_ID))

Jumlah `TapOut_Name` yang belum memiliki `Out_ID`: 19


Karena nilai `TapOut_ID` haruslah unik, kita bisa selanjutnya _generate_ ID baru dengan awalan `T` untuk membedakan sebanyak 19 buah.

In [64]:
new_TapOut_ID = []
for i in range(1, 20):
    new_TapOut_ID.append(f"T{i:05}")
len(new_TapOut_ID)

19

Selanjutnya dipasangkan dengan `TapOut_Name` yang belum memiliki ID

In [65]:
new_TapOut_ID_dict = {}
for x, y in zip(list(no_TapOut_ID), new_TapOut_ID):
    new_TapOut_ID_dict[x] = y

Selanjutnya isi ke DataFrame induk (df) nilai yang baru saja dibuat.

In [66]:
for indeks, data in df.iterrows():
    if data['TapOut_Name'] in new_TapOut_ID_dict.keys():
        df.loc[indeks, 'TapOut_ID'] = new_TapOut_ID_dict[data['TapOut_Name']]

In [67]:
print('Jumlah nilai unik pada kolom `TapOut_ID`  :', df['TapOut_ID'].nunique())
print('Jumlah nilai unik pada kolom `TapOut_Name`:', df['TapOut_Name'].nunique())

Jumlah nilai unik pada kolom `TapOut_ID`  : 2249
Jumlah nilai unik pada kolom `TapOut_Name`: 2249


In [68]:
df[(df['TapOut_ID'].isnull())&(df['TapOut_Name'].notnull())].shape[0]

926

> Terdapat 926 baris `TapOut_Name` yang bernilai notnull dan `TapOut_ID` yang bernilai null.

Selanjunya mengisi nilai `TapOut_ID` yang bernilai null berdasrkan `TapOut_Name`

In [69]:
# Membuat Dataframe yang kolom TapOut_ID dan TapOut_Name yang bernilai notnull
df_TapOut_Name_ID = df[(df['TapOut_ID'].notnull()) & (df['TapOut_Name'].notnull())]

In [70]:
# Membuat Dictionary pasangan TapOut_Name(key) dan TapOut_ID(value)
dict_TapOut_ID_Name = {}
for indeks, data in df_TapOut_Name_ID.iterrows():
    dict_TapOut_ID_Name[data['TapOut_Name']] = data['TapOut_ID']

In [71]:
# Mengisi kolom TapOut_ID yang bernilai Null berdasarkan TapOut_Name
for indeks, data in df.iterrows():
    if pd.notnull(data['TapOut_Name']) and pd.isnull(data['TapOut_ID']):
        df.loc[indeks, 'TapOut_ID'] = dict_TapOut_ID_Name[data['TapOut_Name']]

In [72]:
df.isna().sum()

TransID            0
CardID             0
PayCardBank        0
Name               0
Gender             0
BirthYear          0
CorridorID        88
CorridorName      88
Direction          0
TapIn_ID           0
TapIn_Name         0
TapIn_Lat          0
TapIn_Lon          0
TapIn_Seq          0
TapIn_Time         0
TapOut_ID       1344
TapOut_Name     1344
TapOut_Lat      1344
TapOut_Lon      1344
TapOut_Seq      1344
TapOut_Time     1344
PayAmount       1007
dtype: int64

Selanjutnya mengisi kolom-kolom yang belum memiliki nilai berdasarkan jejak histori dari _customer_, dengan metode yang sama dengan kolom _TapIn_ dan kolom Rute

In [73]:
list_Name_0 = list(df[df['Direction']==0.0]['Name'].unique())
list_Name_1 = list(df[df['Direction']==1.0]['Name'].unique())

In [74]:
# Mengisi nilai null pada kolom TapOut_ID berdasarkan jejak histori customer pada Direction (0.0)
kolom_TapOut = ['TapOut_ID', 'TapOut_Name', 'TapOut_Lat', 'TapOut_Lon', 'TapOut_Seq']

for kolom in kolom_TapOut:
    for nama in list_Name_0:
        dataframe = df[(df['Name']==nama) & (df['Direction']==0.0)]
        if dataframe.shape[0]>2:
            modus_TapIn_ID_0 = dataframe['TapIn_ID'].mode()[0]
            modus_TapIn_Name_0 = dataframe['TapIn_Name'].mode()[0]
            modus_TapIn_Lat_0 = dataframe['TapIn_Lat'].mode()[0]
            modus_TapIn_Lon_0 = dataframe['TapIn_Lon'].mode()[0]
            modus_TapIn_Seq_0 = dataframe['TapIn_Seq'].mode()[0]

            for indeks, data in dataframe.iterrows():
                if pd.isnull(data[kolom]):
                    if data['TapIn_Name'] == modus_TapIn_Name_0:
                        if data['TapIn_Lat'] == modus_TapIn_Lat_0:
                            if data['TapIn_Lon'] == modus_TapIn_Lon_0:
                                if data['TapIn_Seq'] == modus_TapIn_Seq_0:
                                    df.loc[indeks, kolom] = dataframe[kolom].mode()[0]

In [75]:
# Mengisi nilai null pada kolom TapOut_ID berdasarkan jejak histori customer pada Direction (1.0)
for kolom in kolom_TapOut:
    for nama in list_Name_1:
        dataframe = df[(df['Name']==nama) & (df['Direction']==1.0)]
        if dataframe.shape[0]>2:
            modus_TapIn_ID_1 = dataframe['TapIn_ID'].mode()[0]
            modus_TapIn_Name_1 = dataframe['TapIn_Name'].mode()[0]
            modus_TapIn_Lat_1 = dataframe['TapIn_Lat'].mode()[0]
            modus_TapIn_Lon_1 = dataframe['TapIn_Lon'].mode()[0]
            modus_TapIn_Seq_1 = dataframe['TapIn_Seq'].mode()[0]

            for indeks, data in dataframe.iterrows():
                if pd.isnull(data[kolom]):
                    if data['TapIn_Name'] == modus_TapIn_Name_1:
                        if data['TapIn_Lat'] == modus_TapIn_Lat_1:
                            if data['TapIn_Lon'] == modus_TapIn_Lon_1:
                                if data['TapIn_Seq'] == modus_TapIn_Seq_1:
                                    df.loc[indeks, kolom] = dataframe[kolom].mode()[0]

In [76]:
df.isna().sum()

TransID            0
CardID             0
PayCardBank        0
Name               0
Gender             0
BirthYear          0
CorridorID        88
CorridorName      88
Direction          0
TapIn_ID           0
TapIn_Name         0
TapIn_Lat          0
TapIn_Lon          0
TapIn_Seq          0
TapIn_Time         0
TapOut_ID        128
TapOut_Name      128
TapOut_Lat       128
TapOut_Lon       128
TapOut_Seq       128
TapOut_Time     1344
PayAmount       1007
dtype: int64

Tersisa sekitar 128 nilai _missing value_ di beberapa kolom, nilai di baris dan kolom tersebut tidak bisa kita isi karena nilainya acak. Karena tidak bisa memastikan melalui website resmi transjakarta.co.id dengan alasan ada beberapa pembaharuan contohnya ada beberapa CorridorID/CorridorName yang sudah tidak digunakan lagi sehingga tidak bisa mengisi nilai kosong tersebut. Kolom TapOut_Time tidak bisa diisi karena setiap pemberhentian nilainya sulit untuk mendapatkan nilai yang sama, dan kalaupun ada, hal itu merupakan nilai yang random saja. Sehingga kolom-kolom di bagian TapOut yang masih bernilai null tersebut nanti akan dihapus di akhir untuk dilihat akumulasi _missing value_.

Selanjutnya dilakukan cek pada kolom `PayAmount` yang masih bernilai null.

---

### __D. Kolom `PayAmount`__

Pembayaran transjakarta erat kaitannya dengan rute mana jenis layanan apa yang digunakan. Maka dilakukan observasi kolom PayAmount bernilai Null terlebih dahulu.

In [77]:
# Membuat DataFrame `PayAmount` yang bernilai Null
nullpay = df[df['PayAmount'].isna()]
nullpay.head()

TransID            CardID PayCardBank  \
11   IIEK980X1O23NZ   347204308990092      emoney   
76   VHUC320U1S58CB      567286069354       flazz   
97   KFYK655N1H73XS  6590854673757012         dki   
169  IWEH403G6J82OS    30428250357040         bni   
198  WFVC738E7S60SH  3540286623681182         dki   

                            Name Gender  BirthYear CorridorID  \
11         R.M. Labuh Laksmiwati      F       1984        M7B   
76   Dt. Atmaja Suartini, M.Farm      F       2000    JAK.48B   
97             Bajragin Najmudin      M       1986         3H   
169             Estiawan Hastuti      M       2008        M7B   
198      Karimah Pangestu, S.Kom      M       1968    JAK.48A   

                            CorridorName  Direction TapIn_ID  \
11                          BKN - Blok M        1.0   P00254   
76        Stasiun Tebet - Kampung Melayu        0.0  B02272P   
97                       Jelambar - Kota        0.0   P00087   
169                         BKN - Blok M        1.0  B02114P   
198  Stasiun Tebet - Rusun Karet Tengsin        1.0  B05695P   

            TapIn_Name  TapIn_Lat  TapIn_Lon  TapIn_Seq          TapIn_Time  \
11             Tendean  -6.240213  106.83102         12 2023-04-03 06:22:10   
76   Masjid As Sudairi  -6.220626  106.85910          1 2023-04-03 06:12:01   
97            Jelambar  -6.166548  106.78651          0 2023-04-03 06:03:19   
169    Lapangan Blok S  -6.237011  106.81421          6 2023-04-03 05:20:08   
198       Telkom Tebet  -6.224064  106.85160         21 2023-04-03 05:24:09   

    TapOut_ID                      TapOut_Name  TapOut_Lat  TapOut_Lon  \
11     P00129                 Mampang Prapatan   -6.242699   106.82572   
76    B01732P                 Jatinegara Timur   -6.220486   106.86825   
97     P00138                Museum Fatahillah   -6.133930   106.81209   
169   B00290P                          Dammara   -6.254190   106.83146   
198   B02320P  Masjid Jami Al Mamur Bukit Duri   -6.225059   106.85606   

     TapOut_Seq         TapOut_Time  PayAmount  
11         14.0 2023-04-03 06:50:21        NaN  
76         11.0 2023-04-03 06:57:35        NaN  
97          9.0 2023-04-03 06:35:04        NaN  
169        18.0 2023-04-03 06:08:02        NaN  
198        23.0 2023-04-03 05:54:13        NaN

In [78]:
nullpay.groupby('CorridorID').nunique()

TransID  CardID  PayCardBank  Name  Gender  BirthYear  \
CorridorID                                                          
3H               93      15            5    15       2         13   
JAK.48A         255      21            6    21       2         16   
JAK.48B         163      20            6    20       2         17   
M1H             176      20            5    20       2         18   
M7B             320      21            5    21       2         15   

            CorridorName  Direction  TapIn_ID  TapIn_Name  TapIn_Lat  \
CorridorID                                                             
3H                     1          2        11          11         11   
JAK.48A                1          2        24          24         24   
JAK.48B                1          2        14          14         14   
M1H                    1          2        18          18         18   
M7B                    1          2        22          22         22   

            TapIn_Lon  TapIn_Seq  TapIn_Time  TapOut_ID  TapOut_Name  \
CorridorID                                                             
3H                 11          9          93         11           11   
JAK.48A            24         18         255         19           19   
JAK.48B            13         10         163         12           12   
M1H                18         15         176         15           15   
M7B                22         20         320         20           20   

            TapOut_Lat  TapOut_Lon  TapOut_Seq  TapOut_Time  PayAmount  
CorridorID                                                              
3H                  11          11           8           92          0  
JAK.48A             19          19          11          246          0  
JAK.48B             12          12           9          152          0  
M1H                 15          15          11          170          0  
M7B                 20          20          16          308          0

Terdapat 5 RouteID yang unik, maka dilihat karakteristiknya satu-persatu.
- `3H`: Layanan BRT (Bus Rapid Transit)
- `JAK.48A` : Layanan Mikrotrans
- `JAK.48B` : Layanan Mikrotrans
- `M1H` : Belum diketahui
- `M7B` : Belum diketahui

##### 1. CorridorID - `3H`

In [79]:
# membuat dummies untuk mengecek karakteristik pembayaran transjakarta
y=df.dropna()

In [80]:
y[y['CorridorID']=='3H']

Empty DataFrame
Columns: [TransID, CardID, PayCardBank, Name, Gender, BirthYear, CorridorID, CorridorName, Direction, TapIn_ID, TapIn_Name, TapIn_Lat, TapIn_Lon, TapIn_Seq, TapIn_Time, TapOut_ID, TapOut_Name, TapOut_Lat, TapOut_Lon, TapOut_Seq, TapOut_Time, PayAmount]
Index: []

> Seluruh data `PayAmount` di mana nilai `CorridorID` = '3H' adalah null.

In [81]:
R_3H = nullpay[nullpay['CorridorID']=='3H']
R_3H

TransID            CardID PayCardBank  \
97     KFYK655N1H73XS  6590854673757012         dki   
397    YHQJ295K9Z52XX  6590854673757012         dki   
697    ULHQ910M5M77JP  6590854673757012         dki   
997    LJJH685A7R09NL  6590854673757012         dki   
1297   LHKV579I9P41FD  6590854673757012         dki   
...               ...               ...         ...   
33989  KGGP558G8R07NN   213103291564701      emoney   
34176  LFSQ806N1T31FV  3507067369909966         dki   
34381  CZJP861V8W87UN      639030041482       flazz   
34448  DMYZ659H2H38QK     4249227937686      online   
37741  DEMB799H8M93RH      572311836488       flazz   

                              Name Gender  BirthYear CorridorID  \
97               Bajragin Najmudin      M       1986         3H   
397              Bajragin Najmudin      M       1986         3H   
697              Bajragin Najmudin      M       1986         3H   
997              Bajragin Najmudin      M       1986         3H   
1297             Bajragin Najmudin      M       1986         3H   
...                            ...    ...        ...        ...   
33989               Enteng Maryadi      M       1996         3H   
34176          Empluk Wasita, S.T.      F       1973         3H   
34381  Akarsana Nurdiyanti, M.Farm      M       2010         3H   
34448                 Tami Santoso      F       1988         3H   
37741            Cornelia Haryanti      F       1989         3H   

          CorridorName  Direction TapIn_ID                TapIn_Name  \
97     Jelambar - Kota        0.0   P00087                  Jelambar   
397    Jelambar - Kota        1.0   P00294  Sawah Besar Arah Selatan   
697    Jelambar - Kota        0.0   P00087                  Jelambar   
997    Jelambar - Kota        1.0   P00294  Sawah Besar Arah Selatan   
1297   Jelambar - Kota        0.0   P00087                  Jelambar   
...                ...        ...      ...                       ...   
33989  Jelambar - Kota        1.0   P00176                    Petojo   
34176  Jelambar - Kota        0.0   P00291        Harmoni Arah Utara   
34381  Jelambar - Kota        1.0   P00208           RS Sumber Waras   
34448  Jelambar - Kota        0.0   P00098          Kali Besar Barat   
37741  Jelambar - Kota        0.0   P00072                    Glodok   

       TapIn_Lat   TapIn_Lon  TapIn_Seq          TapIn_Time TapOut_ID  \
97     -6.166548  106.786510          0 2023-04-03 06:03:19    P00138   
397    -6.157368  106.818661          4 2023-04-03 17:21:54    P00292   
697    -6.166548  106.786510          0 2023-04-04 06:56:22    P00138   
997    -6.157368  106.818661          4 2023-04-04 17:31:39    P00292   
1297   -6.166548  106.786510          0 2023-04-05 05:40:50    P00138   
...          ...         ...        ...                 ...       ...   
33989  -6.169984  106.816980          6 2023-04-23 18:21:59    P00208   
34176  -6.162678  106.819663          3 2023-04-29 15:26:00    P00293   
34381  -6.166294  106.796750          7 2023-04-30 09:32:06    P00087   
34448  -6.135355  106.811430          8 2023-04-29 17:54:50    P00013   
37741  -6.144500  106.815450          7 2023-04-07 13:19:18    P00013   

                  TapOut_Name  TapOut_Lat  TapOut_Lon  TapOut_Seq  \
97          Museum Fatahillah   -6.133930   106.81209         9.0   
397      Harmoni Arah Selatan   -6.162687   106.81992         5.0   
697         Museum Fatahillah   -6.133930   106.81209         9.0   
997      Harmoni Arah Selatan   -6.162687   106.81992         5.0   
1297        Museum Fatahillah   -6.133930   106.81209         9.0   
...                       ...         ...         ...         ...   
33989         RS Sumber Waras   -6.166294   106.79675         7.0   
34176  Sawah Besar Arah Utara   -6.157418   106.81846         4.0   
34381                Jelambar   -6.166548   106.78651         8.0   
34448                    Kota   -6.136790   106.81490        10.0   
37741                    Kota   -6.136790   106

Berdasarkan layanan transjakarta, Rute `3H` merupakan layanan BRT yang dikenakan tarif Rp3.500. Sehingga nilai yang kosong akan diasumsikan biayanya adalah Rp3.500.

In [82]:
# Mengisi ke DataFrame induk nilai PayAmount
for indeks in R_3H.index:
    df.loc[indeks, 'PayAmount'] = 3500.0

In [83]:
df[df['CorridorID']=='3H']

TransID            CardID PayCardBank  \
97     KFYK655N1H73XS  6590854673757012         dki   
397    YHQJ295K9Z52XX  6590854673757012         dki   
697    ULHQ910M5M77JP  6590854673757012         dki   
997    LJJH685A7R09NL  6590854673757012         dki   
1297   LHKV579I9P41FD  6590854673757012         dki   
...               ...               ...         ...   
33989  KGGP558G8R07NN   213103291564701      emoney   
34176  LFSQ806N1T31FV  3507067369909966         dki   
34381  CZJP861V8W87UN      639030041482       flazz   
34448  DMYZ659H2H38QK     4249227937686      online   
37741  DEMB799H8M93RH      572311836488       flazz   

                              Name Gender  BirthYear CorridorID  \
97               Bajragin Najmudin      M       1986         3H   
397              Bajragin Najmudin      M       1986         3H   
697              Bajragin Najmudin      M       1986         3H   
997              Bajragin Najmudin      M       1986         3H   
1297             Bajragin Najmudin      M       1986         3H   
...                            ...    ...        ...        ...   
33989               Enteng Maryadi      M       1996         3H   
34176          Empluk Wasita, S.T.      F       1973         3H   
34381  Akarsana Nurdiyanti, M.Farm      M       2010         3H   
34448                 Tami Santoso      F       1988         3H   
37741            Cornelia Haryanti      F       1989         3H   

          CorridorName  Direction TapIn_ID                TapIn_Name  \
97     Jelambar - Kota        0.0   P00087                  Jelambar   
397    Jelambar - Kota        1.0   P00294  Sawah Besar Arah Selatan   
697    Jelambar - Kota        0.0   P00087                  Jelambar   
997    Jelambar - Kota        1.0   P00294  Sawah Besar Arah Selatan   
1297   Jelambar - Kota        0.0   P00087                  Jelambar   
...                ...        ...      ...                       ...   
33989  Jelambar - Kota        1.0   P00176                    Petojo   
34176  Jelambar - Kota        0.0   P00291        Harmoni Arah Utara   
34381  Jelambar - Kota        1.0   P00208           RS Sumber Waras   
34448  Jelambar - Kota        0.0   P00098          Kali Besar Barat   
37741  Jelambar - Kota        0.0   P00072                    Glodok   

       TapIn_Lat   TapIn_Lon  TapIn_Seq          TapIn_Time TapOut_ID  \
97     -6.166548  106.786510          0 2023-04-03 06:03:19    P00138   
397    -6.157368  106.818661          4 2023-04-03 17:21:54    P00292   
697    -6.166548  106.786510          0 2023-04-04 06:56:22    P00138   
997    -6.157368  106.818661          4 2023-04-04 17:31:39    P00292   
1297   -6.166548  106.786510          0 2023-04-05 05:40:50    P00138   
...          ...         ...        ...                 ...       ...   
33989  -6.169984  106.816980          6 2023-04-23 18:21:59    P00208   
34176  -6.162678  106.819663          3 2023-04-29 15:26:00    P00293   
34381  -6.166294  106.796750          7 2023-04-30 09:32:06    P00087   
34448  -6.135355  106.811430          8 2023-04-29 17:54:50    P00013   
37741  -6.144500  106.815450          7 2023-04-07 13:19:18    P00013   

                  TapOut_Name  TapOut_Lat  TapOut_Lon  TapOut_Seq  \
97          Museum Fatahillah   -6.133930   106.81209         9.0   
397      Harmoni Arah Selatan   -6.162687   106.81992         5.0   
697         Museum Fatahillah   -6.133930   106.81209         9.0   
997      Harmoni Arah Selatan   -6.162687   106.81992         5.0   
1297        Museum Fatahillah   -6.133930   106.81209         9.0   
...                       ...         ...         ...         ...   
33989         RS Sumber Waras   -6.166294   106.79675         7.0   
34176  Sawah Besar Arah Utara   -6.157418   106.81846         4.0   
34381                Jelambar   -6.166548   106.78651         8.0   
34448                    Kota   -6.136790   106.81490        10.0   
37741                    Kota   -6.136790   106

##### 2. CorridorID - `JAK.48A` & `JAK.48B`

Berdasarkan layanan Transjakarta, rute yang berawalan `JAK...` merupakan layanan mikrotrans.

In [84]:
y[y['CorridorID'].str.contains('JAK')].groupby('PayAmount').nunique()

TransID  CardID  PayCardBank  Name  Gender  BirthYear  CorridorID  \
PayAmount                                                                      
0.0          12984    1040            6  1039       2         64          79   

           CorridorName  Direction  TapIn_ID  TapIn_Name  TapIn_Lat  \
PayAmount                                                             
0.0                  79          2      1439        1439       1434   

           TapIn_Lon  TapIn_Seq  TapIn_Time  TapOut_ID  TapOut_Name  \
PayAmount                                                             
0.0             1380         67       12879       1257         1257   

           TapOut_Lat  TapOut_Lon  TapOut_Seq  TapOut_Time  
PayAmount                                                   
0.0              1253        1206          74        12894

Bila dilihat dari dummies DataFrame diatas dengan kondisi `CorridorID` yang mengandung kata `JAK`, lalu dilakukan _grouping_ berdasarkan `PayAmount`, nilai yang mengandung kata tersebut semua masuk ke dalam kategori pembayaran gratis (0.0). Hal ini didukung berdasarkan layanan transjakarta yang memang layanan mikrotrans merupakan layanan yang gratis. Sehingga kedua rute tersebut dapat diisi dengan angka `0.0`

In [85]:
R_JAK = nullpay[(nullpay['CorridorID']=='JAK.48A') | (nullpay['CorridorID']=='JAK.48B')].index
R_JAK

Index([   76,   198,   219,   376,   498,   519,   676,   798,   819,   976,
       ...
       37365, 37390, 37487, 37511, 37530, 37565, 37590, 37642, 37674, 37882],
      dtype='int64', length=418)

In [86]:
# Mengisi ke DataFrame induk nilai PayAmount
for indeks in R_JAK:
    df.loc[indeks, 'PayAmount'] = 0.0

In [87]:
# Cek kembali ke DataFrame induk
df[(df['CorridorID']=='JAK.48A')|(df['CorridorID']=='JAK.48B')].groupby('PayAmount').nunique()

TransID  CardID  PayCardBank  Name  Gender  BirthYear  CorridorID  \
PayAmount                                                                      
0.0            418      41            6    41       2         27           2   

           CorridorName  Direction  TapIn_ID  TapIn_Name  TapIn_Lat  \
PayAmount                                                             
0.0                   2          2        38          38         38   

           TapIn_Lon  TapIn_Seq  TapIn_Time  TapOut_ID  TapOut_Name  \
PayAmount                                                             
0.0               37         20         418         31           31   

           TapOut_Lat  TapOut_Lon  TapOut_Seq  TapOut_Time  
PayAmount                                                   
0.0                31          31          20          398

##### 3. CorridorID - `M1H` & `M7B`

In [88]:
y[y['CorridorID'].str.contains('M..')].groupby('PayAmount').nunique()

TransID  CardID  PayCardBank  Name  Gender  BirthYear  CorridorID  \
PayAmount                                                                      
3500.0         730      64            6    64       2         33           4   

           CorridorName  Direction  TapIn_ID  TapIn_Name  TapIn_Lat  \
PayAmount                                                             
3500.0                4          2        44          44         44   

           TapIn_Lon  TapIn_Seq  TapIn_Time  TapOut_ID  TapOut_Name  \
PayAmount                                                             
3500.0            44         19         730         44           44   

           TapOut_Lat  TapOut_Lon  TapOut_Seq  TapOut_Time  
PayAmount                                                   
3500.0             44          44          19          730

Dari kode di atas bisa ditarik kesimpulan bahwa `CorridorID` yang berawalan `M` semua melakukan pembayaran sebesar 3500.0. Sehingga dengan asumsi ini, maka _missing value_ pada ID tersebut akan diisi dengan angka 3500.0

In [89]:
R_M = nullpay[(nullpay['CorridorID']== 'M1H') | (nullpay['CorridorID']=='M7B')].index
R_M

Index([   11,   169,   227,   311,   469,   527,   611,   769,   827,   911,
       ...
       36862, 36908, 37062, 37108, 37262, 37308, 37462, 37508, 37661, 37861],
      dtype='int64', length=496)

In [90]:
for indeks in R_M:
    df.loc[indeks, 'PayAmount'] = 3500.0

In [91]:
df.isna().sum()

TransID            0
CardID             0
PayCardBank        0
Name               0
Gender             0
BirthYear          0
CorridorID        88
CorridorName      88
Direction          0
TapIn_ID           0
TapIn_Name         0
TapIn_Lat          0
TapIn_Lon          0
TapIn_Seq          0
TapIn_Time         0
TapOut_ID        128
TapOut_Name      128
TapOut_Lat       128
TapOut_Lon       128
TapOut_Seq       128
TapOut_Time     1344
PayAmount          0
dtype: int64

> Sudah tidak ada `PayAmount` yang bernilai null

---

Setelah semua missing value yang dapat di-handle diisi dengan nilai yang cocok, maka perlu menjalankan kembali kode berikut untuk memisahkan layanan AMARI yang `CorridorName`-nya sama dengan rute utamanya.

In [92]:
list_CorridorID_AMARI = ['M1', 'M2', 'M3', 'M5', 'M9']
list_CorridorName_AMARI = ['Blok M - Kota (M1)', 'Pulo Gadung - Monas (M2)', 'Kalideres - Bundaran HI via Veteran (M3)', 'Matraman Baru - Ancol (M5)', 'Pinang Ranti - Pluit (M9)']

for id, name in zip(list_CorridorID_AMARI, list_CorridorName_AMARI):
    indeks = df[df['CorridorID']==id].index
    for idx in indeks:
        df.loc[idx, 'CorridorName'] = name

---

In [93]:
df.reset_index(drop=True, inplace=True)

In [94]:
df

TransID               CardID PayCardBank  \
0      EIIW227B8L34VB      180062659848800      emoney   
1      LGXO740D2N47GZ     4885331907664776         dki   
2      DJWR385V2U57TO     4996225095064169         dki   
3      JTUZ800U7C86EH         639099174703       flazz   
4      VMLO535V7F95NJ         570928206772       flazz   
...               ...                  ...         ...   
37895  ZWEC949B8Q87QG  4685818286724028395      brizzi   
37896  YHHK837P6Y95GN     6502902290603767         dki   
37897  YXPP627N4G95HO      213159426675861      emoney   
37898  RGVK175U2U98UV      377840859133591      emoney   
37899  FMZZ963S4B68ZP         501862539795       flazz   

                         Name Gender  BirthYear CorridorID  \
0              Bajragin Usada      M       2008          5   
1                Gandi Widodo      F       1997         6C   
2               Emong Wastuti      F       1992        R1A   
3                Surya Wacana      F       1978        11D   
4             Embuh Mardhiyah      M       1982         12   
...                       ...    ...        ...        ...   
37895         Kamila Mahendra      F       2004         6B   
37896            Titi Siregar      M       1974         9N   
37897  drg. Zahra Nashiruddin      F       1976         1T   
37898            Ana Agustina      M       1976     JAK.13   
37899       drg. Leo Najmudin      F       1985         13   

                                   CorridorName  Direction TapIn_ID  \
0                         Matraman Baru - Ancol        1.0   P00142   
1      Stasiun Tebet - Karet via Patra Kuningan        0.0  B01963P   
2                            Pantai Maju - Kota        0.0  B00499P   
3           Pulo Gebang - Pulo Gadung 2 via PIK        0.0  B05587P   
4                         Tanjung Priok - Pluit        0.0   P00239   
...                                         ...        ...      ...   
37895         Ragunan - MH Thamrin via Semanggi        1.0   P00261   
37896                    Pinang Ranti - Pramuka        1.0   P00064   
37897                      Cibubur - Balai Kota        1.0  B02873P   
37898               Tanah Abang - Jembatan Lima        1.0  B02505P   
37899                         Ciledug - Tendean        0.0   P00001   

                   TapIn_Name  TapIn_Lat  TapIn_Lon  TapIn_Seq  \
0                   Pal Putih  -6.184631  106.84402          7   
1                  Kemenkes 2  -6.228700  106.83302         13   
2                Gg. Kunir II  -6.133132  106.81435         38   
3                Taman Elok 1  -6.195743  106.93526         23   
4      Sunter Boulevard Barat  -6.149650  106.88900          5   
...                       ...        ...        ...        ...   
37895                  Tosari  -6.196892  106.82309          2   
37896       Garuda Taman Mini  -6.290154  106.88116          1   
37897           Plaza Sentral  -6.216247  106.81676         12   
37898          Museum Textile  -6.188656  106.80954         33   
37899              Adam Malik  -6.236466  106.74786          2   

               TapIn_Time TapOut_ID            TapOut_Name  TapOut_Lat  \
0     2023-04-03 05:21:44    P00253                Tegalan   -6.203101   
1     2023-04-03 05:42:44   B03307P    Sampoerna Strategic   -6.217152   
2     2023-04-03 05:59:06   B04962P  Simpang Kunir Kemukus   -6.133731   
3     2023-04-03 05:44:51   B03090P      Raya Penggilingan   -6.183068   
4     2023-04-03 06:17:35    P00098       Kali Besar Barat   -6.135355   
...                   ...       ...                    ...         ...   
37895 2023-04-21 18:18:37    P00228                 SMK 57   -6.290967   
37896 2023-04-18 21:52:31    P00179           Pinang Ranti   -6.291075   
37897 2023-04-04 10:29:47   B00226P        Buperta Cibubur   -6.370321   
37898 2023-04-15 19:59:26   B01787P             JPO Blok G   -6.188861   
37899 2023-04-12 21:08:12    P00106         Kebayoran Lama   -6.238340   

       TapOut_Lon  TapOut_Seq        

In [95]:
df_drop = df.dropna()

In [96]:
print('Panjang dataset kotor          :', len(df))
print('Panjang dataset setelah dropna :', len(df_drop), '(clean)')
print('Jumlah baris yang dihapus      :', len(df)-len(df_drop))
print(f'Persentase baris yang dihapus  : {(len(df)-len(df_drop))/len(df)*100:.2f}%')

Panjang dataset kotor          : 37900
Panjang dataset setelah dropna : 36472 (clean)
Jumlah baris yang dihapus      : 1428
Persentase baris yang dihapus  : 3.77%


> Karena persentase jumlah baris yang berisi _missing value_ masih dibawah 5% dan dianggap tidak berpengaruh signifikan, maka untuk kebutuhan analisis akan digunakan dataset setelah menghapus _missing value_ pada dataset aslinya.

In [97]:
df_drop.to_csv('Transjakarta(clean).csv', index=False)